libosmscout/src/osmscout/async/ReadWriteLock.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ This source is part of the libosmscout library
+ Copyright (C) 2024  Jean-Luc Barriere
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
+ */
+
+#include <osmscout/private/Config.h>
+
+#include <osmscout/async/ReadWriteLock.h>
+
+#include <osmscout/log/Logger.h>
+
+#include <cassert>
+
+namespace osmscout {
+
+/**
+ * X flag is defined as the following:
+ * 0 : X is released
+ * 1 : X is held, but waits for release of S
+ * 2 : X is held
+ * 3 : X was released and left available for one of request in wait
```

Comment:
I would suggest to use named contants (constexpr) for these meanings...

Reply:
mmm, using addl vs mov is better for performance. Also x_flag is atomic, and I use as much as possible "lock add/sub 1"

Reply:
I would just replace numeric constants like: `if (x_flag == 0)` with:
```
constexpr int X_RELEASED = 0;
...
if (x_flag == X_RELEASED)
```
generated assembler will be the same